crates/runtime/src/embeddings/index/s3/index.rs

```diff
@@ -132,7 +146,68 @@ impl SearchIndex for S3Vector {
         &self,
         record: RecordBatch,
     ) -> Result<RecordBatch, Box<dyn std::error::Error + Send + Sync>> {
-        write::write(self, record).await.boxed()
+        match self.partition_by.first() {
+            Some(partition_by) => {
+                let input_dfschema = DFSchema::try_from(record.schema())?;
+                let execution_props = ExecutionProps::new();
+                let physical_expr =
+                    create_physical_expr(partition_by, &input_dfschema, &execution_props)?;
+                let partitions = partition_batch(&record, physical_expr.as_ref())?;
+
+                for (partition_value, partition_record) in partitions.into_values() {
+                    // change the index name to a partition name
+                    let id = match &self.table.idx {
+                        S3VectorIdentifier::IndexArn(_) => {
+                            tracing::warn!(
+                                "Partitioning is not supported when index ARN is provided. Please provide the bucket and index name instead."
+                            );
+                            return write::write(self, &self.table, record).await.boxed();
+                        }
+                        S3VectorIdentifier::Index {
+                            bucket_name,
+                            index_name,
+                        } => {
+                            let partitioned_index_name = PartitionedIndexName::new(
+                                index_name,
+                                &self.embedded_column,
+                                &partition_value,
+                                &self.partition_by,
+                            )?;
+                            let index_name = partitioned_index_name.to_index_name();
+                            tracing::trace!(
+                                "writing {} records to index: {index_name}",
+                                partition_record.num_rows(),
+                            );
+                            S3VectorIdentifier::Index {
+                                bucket_name: bucket_name.clone(),
+                                index_name,
+                            }
+                        }
+                    };
+
+                    let table = S3VectorsTable::try_create_new_table(
```

Comment:
I think if we used `.try_new_table(` and not `.try_create_new_table(`, we wouldn't need to store `dimensions`, `columns` or `distance_metric`?

Reply:
But we need to create the index if it doesn't exist. `try_create_new_table` does that. `try_new_table` does not.

Reply:
Sorry, yes correct.